src/ring_theory/witt_vector/isocrystal.lean

```diff
@@ -0,0 +1,201 @@
+
+/-- A helper instance for type class inference. -/
+local attribute [instance]
+def fraction_ring.module : module K(p, k) K(p, k) := semiring.to_module
+
+/--
+Type synonym for `K(p, k)` to carry the standard `m`-indexed 1-dimensional isocrystal structure.
+-/
+@[nolint unused_arguments has_inhabited_instance, derive [add_comm_group, module K(p, k)]]
+def standard_one_dim_isocrystal (m : ℤ) : Type* :=
+K(p, k)
+
+section perfect_ring
+variables [is_domain k] [char_p k p] [perfect_ring k p]
+
+/-- The standard one-dimensional isocrystal is an isocrystal. -/
+instance (m : ℤ) : isocrystal p k (standard_one_dim_isocrystal p k m) :=
+{ frob := (fraction_ring.frobenius p k).to_semilinear_equiv.trans
+   (linear_equiv.smul_of_ne_zero _ _ _ (zpow_ne_zero m (witt_vector.fraction_ring.p_nonzero p k))) }
+
+@[simp] lemma standard_one_dim_isocrystal.frobenius_apply (m : ℤ)
+  (x : standard_one_dim_isocrystal p k m) :
+  Φ(p, k) x = (p:K(p, k)) ^ m • φ(p, k) x :=
+rfl
+
+/-- A one-dimensional isocrystal admits an isomorphism to one of the standard (indexed by `m : ℤ`)
+one-dimensional isocrystals. -/
+theorem isocrystal_classification
+  (k : Type*) [field k] [is_alg_closed k] [char_p k p]
+  (V : Type*) [add_comm_group V] [isocrystal p k V]
+  (h_dim : finrank K(p, k) V = 1) :
+  ∃ (m : ℤ), nonempty (standard_one_dim_isocrystal p k m ≃ᶠⁱ[p, k] V) :=
+begin
+  haveI : nontrivial V := finite_dimensional.nontrivial_of_finrank_eq_succ h_dim,
+  obtain ⟨x, hx⟩ : ∃ x : V, x ≠ 0 := exists_ne 0,
+  have : Φ(p, k) x ≠ 0 := by simpa using Φ(p, k).injective.ne hx,
+  obtain ⟨a, ha, hax⟩ : ∃ a : K(p, k), a ≠ 0 ∧ Φ(p, k) x = a • x,
+  { rw finrank_eq_one_iff_of_nonzero' x hx at h_dim,
+    obtain ⟨a, ha⟩ := h_dim (Φ(p, k) x),
+    refine ⟨a, _, ha.symm⟩,
+    intros ha',
+    apply this,
+    simp [← ha, ha'] },
+  obtain ⟨b, hb, m, (hmb : φ(p, k) b * a = p ^ m * b)⟩ :=
+    witt_vector.exists_frobenius_solution_fraction_ring p ha,
+  use m,
+  let F₀ : standard_one_dim_isocrystal p k m →ₗ[K(p,k)] V :=
+    linear_map.to_span_singleton K(p, k) V x,
+  let F : standard_one_dim_isocrystal p k m ≃ₗ[K(p,k)] V,
+  { refine linear_equiv.of_bijective F₀ _ _,
+    { rw ← linear_map.ker_eq_bot,
+      exact linear_equiv.ker_to_span_singleton K(p, k) V hx },
+    { rw ← linear_map.range_eq_top,
+      rw ← (finrank_eq_one_iff_of_nonzero x hx).mp h_dim,
+      rw linear_map.span_singleton_eq_range } },
+  refine ⟨⟨(linear_equiv.smul_of_ne_zero K(p, k) _ _ hb).trans F, _⟩⟩,
+  intros c,
+  simp only [hax, ←mul_smul, linear_equiv.trans_apply, linear_equiv.smul_of_ne_zero_apply,
+    algebra.id.smul_eq_mul, linear_equiv.of_bijective_apply, linear_map.to_span_singleton_apply,
+    linear_equiv.map_smulₛₗ, ring_equiv.coe_to_ring_hom, ring_equiv.map_mul, ring_hom.map_mul,
+    standard_one_dim_isocrystal.frobenius_apply],
+  congr' 1,
+  transitivity (φ(p,k) b * a) * φ(p,k) c,
+  { ring, },
+  { rw hmb,
+    ring },
+end
+
+end perfect_ring
```

Comment:
Did you mean to `end perfect_ring` above this lemma and then introduce
```lean
variables (k : Type*) [field k] [is_alg_closed k] [char_p k p]
```
for the isocrystal classification theorem?